File: api/app/languages.py
"""
Languages, and what a learner's first language is allowed to affect.

PhonoPlay can personalize practice for learners whose first language differs
from their target language. The whole of that personalization happens in
three places, and it is worth being precise about which:

    1. the interface       — names, scripts, and which sounds are offered
    2. the bridge          — a familiar sound used as a starting point
    3. the material prompt — context given to the exercise generator

It affects **nothing else**. In particular the first language is never an
input to the acoustic measurement. A learner practising English /θ/ is
measured against the English /θ/ reference profile whoever they are, and
`acoustic.analyze()` has no parameter through which a language could reach
it. That is deliberate on two counts:

  * **It keeps the measurement stable.** Adding a language cannot change what
    an existing recording scores.
  * **It is the only honest option.** The reference profiles are built from
    English audio (see `acoustic/reference/README.md`). We have no Bangla
    acoustic data, so Bangla can be a *native* language here but not a
    *target* one, and the registry below says so rather than pretending
    otherwise.

## On causation

Nothing in this module claims that a learner's first language causes any
pronunciation pattern. That claim is not ours to make: it would require
evidence about individual learners that a two-second recording cannot supply,
and stated carelessly it turns a description of a language into a prediction
about a person.

What the bridges below *do* say is narrower and checkable: that a particular
sound in one language is articulated in a particular place, and that a
particular sound in another language shares that place. "থ puts the tongue
tip at the teeth, and so does /θ/" is a description of two articulations. It
is useful precisely because it gives a learner somewhere familiar to start —
not because it predicts anything about what they will do.

Not to be confused with `app/stt/languages.py`, which is an internal helper
of the transcription package that maps provider language names to ISO-639-1
codes. This module is the product-level registry: which languages a learner
can choose, and what choosing one is allowed to change.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class Language:
    code: str
    #: English name, for English-language interface copy.
    name: str
    #: The language's own name in its own script.
    native_name: str
    script: str
    #: Whether PhonoPlay can *measure* pronunciation in this language. False
    #: means "no acoustic reference data", not "less important".
    can_be_target: bool
    #: Whether it can be selected as a learner's first language.
    can_be_native: bool
    #: Shown in the picker when `can_be_target` is False, so the limitation is
    #: visible at the point it matters rather than buried in a doc.
    target_note: str | None = None


LANGUAGES: dict[str, Language] = {
    "en": Language(
        code="en",
        name="English",
        native_name="English",
        script="Latin",
        can_be_target=True,
        can_be_native=True,
    ),
    "bn": Language(
        code="bn",
        name="Bangla",
        native_name="বাংলা",
        script="Bengali",
        # PhonoPlay measures English sounds against English reference audio.
        # There is no Bangla reference data, so there is nothing to measure a
        # Bangla target against, and inventing one would be the exact
        # fabrication this project refuses.
        can_be_target=False,
        can_be_native=True,
        target_note=(
            "PhonoPlay does not measure Bangla pronunciation yet — the "
            "acoustic reference data is English only. Bangla can be your "
            "first language here, and practice is personalized for it."
        ),
    ),
}

DEFAULT_NATIVE = "en"
DEFAULT_TARGET = "en"


@dataclass(frozen=True)
class BridgeStep:
    """One step on the way from a familiar sound to the target."""

    #: What the learner sees. A grapheme, an IPA symbol, or a word.
    text: str
    #: "native" | "target" | "word"
    kind: str
    #: IPA for the sound this step is about, where there is a single one.
    ipa: str | None = None
    #: One line about what changes at this step.
    note: str | None = None


@dataclass(frozen=True)
class Bridge:
    """
    A familiar starting point for one target sound, for one first language.

    `anchor_note` describes an articulatory relationship between two sounds.
    It is not a claim about what any learner does or will do.
    """

    native: str
    target_sound: str
    #: The familiar grapheme, or None when the learner's first language is
    #: also the target language and there is nothing to bridge from.
    anchor: str | None
    anchor_ipa: str | None
    anchor_note: str | None
    steps: list[BridgeStep] = field(default_factory=list)

    @property
    def progression(self) -> list[str]:
        return [step.text for step in self.steps]


def _words(*pairs: tuple[str, str]) -> list[BridgeStep]:
    return [BridgeStep(text=text, kind="word", note=note) for text, note in pairs]


#: Bangla-to-English bridges.
#:
#: Every anchor below is a **shared place of articulation**, stated as a
#: description of two articulations rather than a prediction about a speaker.
#: Bangla's dental stop series (ত থ দ ধ) is laminal dental, the same region
#: English /θ/ uses; ল and English /l/ both make alveolar contact; র is an
#: alveolar tap or trill in the same region English /ɹ/ approaches.
#:
#: Every word in every progression begins with the target sound, which the
#: journey requires: the acoustic stage measures the onset of the utterance
#: (ARCHITECTURE.md §15), so a progression word that did not start with the
#: sound would be unmeasurable.
_BN_EN: dict[str, Bridge] = {
    "th": Bridge(
        native="bn",
        target_sound="th",
        anchor="থ",
        anchor_ipa="t̪ʰ",
        anchor_note=(
            "থ already puts your tongue tip at your teeth — the same place "
            "English /θ/ uses. The difference is what the air does: থ stops "
            "it, /θ/ keeps it flowing."
        ),
        steps=[
            BridgeStep("থ", "native", "t̪ʰ", "The place you already know."),
            BridgeStep("θ", "target", "θ", "Same place. Let the air keep going."),
            *_words(
                ("think", "The sound at the start of a word."),
                ("three", "Now with /r/ straight after it."),
                ("through", "The same cluster, a different vowel."),
            ),
        ],
    ),
    "s": Bridge(
        native="bn",
        target_sound="s",
        anchor="স",
        anchor_ipa="s ~ ʃ",
        anchor_note=(
            "স is written for a sibilant, and across Bangla varieties it is "
            "produced as [s] or as [ʃ]. English /s/ is consistently the "
            "higher, sharper of the two."
        ),
        steps=[
            BridgeStep("স", "native", "s ~ ʃ", "A sibilant you already make."),
            BridgeStep("s", "target", "s", "Narrow the groove; aim the air higher."),
            *_words(
                ("sun", "One syllable, sound at the start."),
                ("snake", "Now with a consonant straight after."),
                ("street", "A three-consonant cluster."),
            ),
        ],
    ),
    "r": Bridge(
        native="bn",
        target_sound="r",
        anchor="র",
        anchor_ipa="ɾ ~ r",
        anchor_note=(
            "র taps or trills the tongue against the ridge behind your top "
            "teeth. English /ɹ/ moves toward that same region but never "
            "makes contact."
        ),
        steps=[
            BridgeStep("র", "native", "ɾ ~ r", "The region you already use."),
            BridgeStep("ɹ", "target", "ɹ", "Approach, but do not touch."),
            *_words(
                ("red", "One syllable, sound at the start."),
                ("rain", "Hold the /r/ before the vowel."),
                ("rocket", "Two syllables, same opening sound."),
            ),
        ],
    ),
    "l": Bridge(
        native="bn",
        target_sound="l",
        anchor="ল",
        anchor_ipa="l",
        anchor_note=(
            "ল and English /l/ are both made with the tongue tip touching "
            "the ridge behind the top teeth. This is the closest pairing of "
            "the four."
        ),
        steps=[
            BridgeStep("ল", "native", "l", "Very close to the English sound."),
            BridgeStep("l", "target", "l", "Same contact, held a little longer."),
            *_words(
                ("light", "One syllable, sound at the start."),
                ("leaf", "Keep the contact firm before the vowel."),
                ("lemonade", "Three syllables, same opening sound."),
            ),
        ],
    ),
}

#: English-to-English: no cross-language anchor, so the progression starts at
#: the target sound itself. English-only mode is not a degraded path — it is
#: simply the case where there is nothing to bridge from.
_EN_EN: dict[str, Bridge] = {
    "th": Bridge("en", "th", None, None, None, [
        BridgeStep("θ", "target", "θ", "The sound on its own."),
        *_words(("think", "At the start of a word."),
                ("three", "Now with /r/ straight after it."),
                ("through", "The same cluster, a different vowel.")),
    ]),
    "s": Bridge("en", "s", None, None, None, [
        BridgeStep("s", "target", "s", "The sound on its own."),
        *_words(("sun", "At the start of a word."),
                ("snake", "With a consonant straight after."),
                ("street", "A three-consonant cluster.")),
    ]),
    "r": Bridge("en", "r", None, None, None, [
        BridgeStep("ɹ", "target", "ɹ", "The sound on its own."),
        *_words(("red", "At the start of a word."),
                ("rain", "Hold it before the vowel."),
                ("rocket", "Two syllables.")),
    ]),
    "l": Bridge("en", "l", None, None, None, [
        BridgeStep("l", "target", "l", "The sound on its own."),
        *_words(("light", "At the start of a word."),
                ("leaf", "Keep the contact firm."),
                ("lemonade", "Three syllables.")),
    ]),
}

_BRIDGES: dict[str, dict[str, Bridge]] = {"bn": _BN_EN, "en": _EN_EN}


def is_language(code: str) -> bool:
    return code in LANGUAGES


def language(code: str) -> Language:
    try:
        return LANGUAGES[code]
    except KeyError:
        raise ValueError(f"unknown language {code!r}") from None


def native_options() -> list[Language]:
    return [lang for lang in LANGUAGES.values() if lang.can_be_native]


def target_options() -> list[Language]:
    return [lang for lang in LANGUAGES.values() if lang.can_be_target]


def normalise_native(code: str | None) -> str:
    """Fall back to the default rather than raising on unknown input."""
    if code and code in LANGUAGES and LANGUAGES[code].can_be_native:
        return code
    return DEFAULT_NATIVE


def bridge_for(native: str, target_sound: str) -> Bridge | None:
    """
    The bridge for one sound, or None when there is no entry.

    Returning None is a normal outcome: a language pair with no researched
    bridge gets the plain target-language progression rather than an invented
    one.
    """
    return _BRIDGES.get(normalise_native(native), {}).get(target_sound)


def as_dict(bridge: Bridge | None) -> dict | None:
    if bridge is None:
        return None
    return {
        "native": bridge.native,
        "target_sound": bridge.target_sound,
        "anchor": bridge.anchor,
        "anchor_ipa": bridge.anchor_ipa,
        "anchor_note": bridge.anchor_note,
        "progression": bridge.progression,
        "steps": [
            {"text": s.text, "kind": s.kind, "ipa": s.ipa, "note": s.note}
            for s in bridge.steps
        ],
    }


def language_dict(code: str) -> dict:
    lang = language(code)
    return {
        "code": lang.code,
        "name": lang.name,
        "native_name": lang.native_name,
        "script": lang.script,
        "can_be_target": lang.can_be_target,
        "can_be_native": lang.can_be_native,
        "target_note": lang.target_note,
    }

File: web/src/language/languages.ts
import { ENGLISH_PHONEMES } from './english.ts'
import type { LanguageCode, LanguagePairProfile, LanguageProfile } from './types.ts'

/**
 * Languages, and what pairing two of them is allowed to imply.
 *
 * ## Two languages, honestly
 *
 * English and Bangla. Bangla can be a learner's first language but not a
 * target, because PhonoPlay has no acoustic reference data for Bangla. That
 * is a statement about our data, not about the language, and the picker says
 * so at the point the choice is made.
 *
 * ## What a language pair does NOT say
 *
 * The hints below order the **baseline assessment** so a learner meets the
 * most informative sounds first. They are not predictions.
 *
 * In particular, the fact that a sound is absent from a language's inventory
 * is not evidence that a speaker of it will struggle with that sound. Plenty
 * of people produce sounds their first language does not contain, and a
 * learner's actual recording is the only thing that settles it. Every hint is
 * therefore phrased as a description of two sound systems, and every one is
 * discarded as soon as a measurement exists.
 */

export const LANGUAGES: Record<LanguageCode, LanguageProfile> = {
  en: {
    code: 'en',
    name: 'English',
    nativeName: 'English',
    script: 'Latin',
    direction: 'ltr',
    canBeNative: true,
    canBeTarget: true,
    phonemes: ENGLISH_PHONEMES.map((p) => p.id),
  },
  es: {
    code: 'es',
    name: 'Spanish',
    nativeName: 'Español',
    script: 'Latin',
    direction: 'ltr',
    canBeNative: true,
    canBeTarget: false,
    targetNote:
      'PhonoPlay does not measure Spanish pronunciation yet — the acoustic ' +
      'reference data is English only. Spanish can be your first language ' +
      'here.',
    // Selectable as a first language, but there is no researched
    // Spanish-to-English bridge yet, so `getLanguagePairProfile('es', 'en')`
    // returns undefined and the assessment falls back to the plain English
    // ordering. That is the honest behaviour: an invented anchor would be
    // worse than none.
    phonemes: [],
  },
  bn: {
    code: 'bn',
    name: 'Bangla',
    nativeName: 'বাংলা',
    script: 'Bengali',
    direction: 'ltr',
    canBeNative: true,
    // No Bangla reference audio exists, so there is nothing to measure a
    // Bangla target against. Inventing one would be fabrication.
    canBeTarget: false,
    targetNote:
      'PhonoPlay does not measure Bangla pronunciation yet — the acoustic ' +
      'reference data is English only. Bangla can be your first language ' +
      'here, and practice is personalized for it.',
    // Bangla contributes anchors rather than practice targets, so it lists no
    // phonemes of its own in the knowledge base yet.
    phonemes: [],
  },
}

/**
 * The caveat that travels with every pair profile.
 *
 * Carried on the data rather than left to whichever component renders it to
 * remember to say.
 */
const OVERRIDE_NOTE =
  'These are starting points for what to assess first, not predictions about ' +
  'any learner. Measured results replace them.'

/**
 * Bangla → English.
 *
 * Each anchor is a **shared place of articulation** — a description of two
 * articulations, chosen because starting from a sound someone already makes
 * is better teaching than starting from nothing.
 *
 * Bangla's dental stop series (ত থ দ ধ) is laminal dental, the same region
 * English /θ/ uses. ল and English /l/ both make alveolar contact. র is an
 * alveolar tap or trill in the region English /ɹ/ approaches without
 * touching.
 */
const BN_EN: LanguagePairProfile = {
  native: 'bn',
  target: 'en',
  crossLanguage: true,
  hints: [
    {
      phoneme: 'th',
      suggestedOrder: 1,
      rationale:
        'Bangla has a dental stop series (ত থ দ ধ) made at the same place ' +
        'as English /θ/. Assessing it early is informative because the ' +
        'place is already familiar and only the manner differs — a stop ' +
        'closes the airflow, a fricative keeps it going.',
      anchor: {
        grapheme: 'থ',
        ipa: 't̪ʰ',
        note:
          'থ already puts your tongue tip at your teeth — the same place ' +
          'English /θ/ uses. The difference is what the air does: থ stops ' +
          'it, /θ/ keeps it flowing.',
      },
    },
    {
      phoneme: 'r',
      suggestedOrder: 2,
      rationale:
        'র is an alveolar tap or trill; English /ɹ/ approaches the same ' +
        'region without making contact. Same neighbourhood, different manner.',
      anchor: {
        grapheme: 'র',
        ipa: 'ɾ ~ r',
        note:
          'র taps or trills the tongue against the ridge behind your top ' +
          'teeth. English /r/ moves toward that same region but never ' +
          'touches.',
      },
    },
    {
      phoneme: 's',
      suggestedOrder: 3,
      rationale:
        'স is written for a sibilant, and across Bangla varieties it is ' +
        'produced as [s] or as [ʃ]. Assessing it establishes which one this ' +
        'learner is producing rather than assuming.',
      anchor: {
        grapheme: 'স',
        ipa: 's ~ ʃ',
        note:
          'স is a sibilant you already make. English /s/ is consistently ' +
          'the higher, sharper of the two.',
      },
    },
    {
      phoneme: 'l',
      suggestedOrder: 4,
      rationale:
        'ল and English /l/ are both alveolar laterals. This is the closest ' +
        'pairing of the four, so it is assessed last — it is the least ' +
        'likely to be informative early on.',
      anchor: {
        grapheme: 'ল',
        ipa: 'l',
        note:
          'ল and English /l/ are both made with the tongue tip touching the ' +
          'ridge behind the top teeth.',
      },
    },
  ],
  note: OVERRIDE_NOTE,
}

/**
 * English → English.
 *
 * No cross-language anchor, because there is nothing to bridge from. Ordered
 * by the general difficulty of the sounds in English, which is a property of
 * the sounds rather than of the speaker.
 *
 * This is not a degraded path. It is simply the case where the learner's
 * first language is the one they are practising.
 */
const EN_EN: LanguagePairProfile = {
  native: 'en',
  target: 'en',
  crossLanguage: false,
  hints: [
    {
      phoneme: 'th',
      suggestedOrder: 1,
      rationale:
        '/θ/ is among the later sounds to settle in English and is easily ' +
        'confused with /f/ and /s/, so it is informative to assess first.',
    },
    {
      phoneme: 'r',
      suggestedOrder: 2,
      rationale:
        'English /ɹ/ is articulated without contact, which makes it unusual ' +
        'and worth establishing early.',
    },
    {
      phoneme: 's',
      suggestedOrder: 3,
      rationale:
        '/s/ is loud and well-defined acoustically, so a reading of it is ' +
        'reliable and gives a good reference point for the others.',
    },
    {
      phoneme: 'l',
      suggestedOrder: 4,
      rationale:
        '/l/ has clear tongue contact and is generally the most stable of ' +
        'the four.',
    },
  ],
  note: OVERRIDE_NOTE,
}

/** Keyed `native>target`. */
export const LANGUAGE_PAIRS: Record<string, LanguagePairProfile> = {
  'bn>en': BN_EN,
  'en>en': EN_EN,
}

export function pairKey(native: LanguageCode, target: LanguageCode): string {
  return `${native}>${target}`
}
